### PR TITLE
Add subcell boundary conditions to GRMHD

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes

Enable using subcell at external boundaries for GRMHD system.
cf. See #3746 for the same feature for Burgers.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp
Comment (yoonso0-0) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp]: Is there a way to allocate two `Variables` (with the one on the line 197) at once to reduce overhead?

Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp]: Not in one line, but my non-owning Variables PR will make this possible. It'll be 3 lines, though (probably worth it :) )

Comment (yoonso0-0) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp]: Okay, my eyes on that PR.

**Comment by yoonso0-0** [comment]
Rebased and made small fixes from Larry's review on Burgers PR.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp
Comment (yoonso0-0) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp:176]: Not sure this is the best way to count the number of individual components of `<prim_tags_for_reconstruction>`..

**Comment by nilsdeppe** [comment]
@yoonso0-0 this is still failing CI

**Comment by yoonso0-0** [comment]
I closed a separate PR for subcell FreeOutflow boundary condition, but left it as a separate commit here to make review easier. I'll squash the FD FreeOutflow commit into FD boundary condition one once review's finished.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp]: Can you allocate this vector before the `Variables` and then use a non-owning Variables? That'll eliminate a copy+allocation

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp]: I think here we want `std::move(boundary_ghost_data)` to eliminate another copy+allocation

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/ReconstructWork.tpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/ReconstructWork.tpp:209]: maybe add an `ASSERT(element.external_boundaries().count(direction) == 1, "...");`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Looks good, few minor suggestions. Please rebase and squash immediately :)

**Comment by yoonso0-0** [comment]
Squashed fixups into "Add FD boundary condition to GRMHD subcell" and rebased. I'll squash last two commits once approved by another review.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp
Comment (yoonso0-0) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp]: Forgot to update these lines after we got the new feature :b

**Review by nilsdeppe — COMMENTED** [review]
LGTM. Given how long this PR's been open I wouldn't wait for a second reviewer. It's not required to have 2 reviewers

**Comment by yoonso0-0** [comment]
@nilsdeppe Okay!

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
